{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_FOLDER = '../../ADA2017-Tutorials/02 - Intro to Pandas/Data' # Use the data folder provided in Tutorial 02 - Intro to Pandas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# config\n",
    "%matplotlib inline\n",
    "\n",
    "# imports\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "from datetime import datetime\n",
    "import calendar\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Methodology\n",
    "\n",
    "We start by manually browsing through the data to get an understanding of the contents and identify the variables of interest. We quickly realize that there are naming inconsistencies, formatting issues, as well as missing and potentially invalid values.\n",
    "\n",
    "We opt to use cumulative values instead of daily values for two reasons:\n",
    "1. As opposed to daily values, cumulative totals are present in all three datasets.\n",
    "2. Cumulative values are more tolerant of missing values, since they cumulate all previous days.\n",
    "3. It seems easier to correct invalid cumulative values because they are supposed to be monotonic increasing \n",
    "\n",
    "We parse in all the CSV files per country (using `pandas.read_csv`) and concatenate them together into a single frame. We perform cleanup on a per-country basis as the CSV labels are different across countries and each country requires slightly different cleanup rules, which describe below. The cleanup phase also extracts the relevant data and produces standardized dataframes with the following structure: $(date, value)$. We collect two such dataframes per country: one for the new reported cases and one for the deaths.\n",
    "\n",
    "After collecting cumulative series, we need to clean up the data and convert the results into daily values in order to compute the requested averages.\n",
    "\n",
    "We clean up a cumulative series using the utility function `clean_frame` below. This function computes a range of valid values using the min and max values, excludes values that fall outside this range, and then drops all values which do not increase monotonically. Once we have a clean cumulative series, we compute the corresponding daily series using the `cum_to_daily` function, which simply subtracts consecutive values in the series (we transpose the frame with `T` and then use `diff`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Utility functions\n",
    "\n",
    "def clean_frame(df):\n",
    "    # Drop values which are not in the interval between first and last values\n",
    "    minv = df.value[0]\n",
    "    maxv = df.value[-1]\n",
    "    df = df.query('value >= %d and value <= %d' % (minv, maxv)).copy()\n",
    "\n",
    "    # Drop values which do not fit the monotonic criteria\n",
    "    i = 0\n",
    "    for index, row in df.iterrows():\n",
    "        if i <= row.value:\n",
    "            i = row.value\n",
    "        else:\n",
    "            df.drop(index, inplace=True)\n",
    "    \n",
    "    # Let's make sure it's monotonic now\n",
    "    assert((df['value'].T.diff().fillna(0) >= 0).all())\n",
    "    return df\n",
    "\n",
    "def cum_to_daily(df):\n",
    "    df = df.copy()\n",
    "    df = df.value.set_value(min(df.index) - pd.DateOffset(1), 0).sort_index() # prepend a 0 to capture the first daily value\n",
    "    return df.T.diff()[1:]\n",
    "\n",
    "def country_frame(df1, df2, country):\n",
    "    merged = df1.merge(df2, how='outer', left_index=True, right_index=True).fillna(0)\n",
    "    merged['country'] = country\n",
    "    return merged.set_index(['country'], append=True).swaplevel()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Guinea CSVs appear to be the cleanest input files and have consistent naming (at least for the values we are interested in). Formatting-wise, the main issue is the storing of values as `string`s, which we converted to numeric values using `pandas.to_numeric`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 426,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in Guinea CSV\n",
    "guinea_df = pd.concat((pd.read_csv(f, usecols=['Date', 'Description', 'Totals'], parse_dates=['Date']) \n",
    "                       for f in glob.glob(DATA_FOLDER+'/ebola/guinea_data/*.csv')))\n",
    "guinea_df.columns = ['date', 'key', 'value']\n",
    "guinea_df[['value']] = guinea_df[['value']].apply(pd.to_numeric, errors='coerce')\n",
    "\n",
    "# Collect new cases\n",
    "guinea_cases = guinea_df.loc[guinea_df.key == 'Cumulative (confirmed + probable + suspects)'][['date', 'value']].set_index('date').sort_index()\n",
    "\n",
    "# Collect deaths\n",
    "guinea_deaths = guinea_df.loc[guinea_df.key == 'Total deaths (confirmed + probables + suspects)'][['date', 'value']].set_index('date').sort_index()\n",
    "\n",
    "# Clean up the two frames\n",
    "guinea_cases_cum = clean_frame(guinea_cases)\n",
    "guinea_deaths_cum = clean_frame(guinea_deaths)\n",
    "\n",
    "# Convert to daily\n",
    "guinea_cases_daily = cum_to_daily(guinea_cases).to_frame('cases')\n",
    "guinea_deaths_daily =  cum_to_daily(guinea_deaths).to_frame('deaths')\n",
    "\n",
    "# Combine the two into a single frame\n",
    "guinea = country_frame(guinea_cases_daily, guinea_deaths_daily, 'Guinea')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Liberia CSVs are a little more trickier.\n",
    "* First, in terms of formatting, the same variable name can appear in multiple variants, sometimes with new lines in the name. We use a simple regular expression to capture the different cases.\n",
    "* Second, in terms of content, we observed that there are many blanks, so looking at a single variable for the total number of cases and deaths is not good enough. Instead, we look at both the aggregate (probable + suspected + confirmed) and the individual values: we collect both and if they do not match or some values are missing, wecombine them into a final value per date using the maximum of the two. The choice of `max()` is arbitrary, and may lead to a pessimistic estimate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Read in Liberia CSV\n",
    "liberia_df = pd.concat((pd.read_csv(f, usecols=['Date', 'Variable', 'National'], parse_dates=['Date'])\n",
    "                        for f in glob.glob(DATA_FOLDER+'/ebola/liberia_data/*.csv')))\n",
    "liberia_df.columns = ['date', 'key', 'value']\n",
    "\n",
    "# Collect new cases\n",
    "df = liberia_df.copy()\n",
    "df1 = df.loc[df.key.str.match('.*Cumulative.*confirmed.*probable.*suspected.*', flags=re.DOTALL)][['date', 'value']].set_index('date')\n",
    "df2 = df.loc[df.key.isin(['Total suspected cases', 'Total probable cases', 'Total confirmed cases'])].groupby('date').sum()\n",
    "liberia_cases = pd.concat([df1, df2]).sort_index().groupby('date').max().dropna()\n",
    "\n",
    "# Collect deaths\n",
    "df1 = df.loc[df.key.str.match('.*Total.*death.*confirmed.*probable.*suspected.*', flags=re.DOTALL)][['date', 'value']].set_index('date')\n",
    "df2 = df.loc[df.key.isin(['Total death/s in suspected cases', 'Total death/s in probable cases', 'Total death/s in confirmed cases'])].groupby('date').sum()\n",
    "liberia_deaths = pd.concat([df1, df2]).sort_index().groupby('date').max().dropna()\n",
    "\n",
    "# Clean up the two frames\n",
    "liberia_cases_cum = clean_frame(liberia_cases)\n",
    "liberia_deaths_cum = clean_frame(liberia_deaths)\n",
    "\n",
    "# Convert to daily\n",
    "liberia_cases_daily = cum_to_daily(liberia_cases).to_frame('cases')\n",
    "liberia_deaths_daily =  cum_to_daily(liberia_deaths).to_frame('deaths')\n",
    "\n",
    "# Combine the two into a single frame\n",
    "liberia = country_frame(liberia_cases_daily, liberia_deaths_daily, 'Liberia')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sierra Leone CSVs require a bit more processing.\n",
    "* Formatting-wise, there are no surprises, other than the fact that some values as stored as `string`s, which we convert to numeric values as we did in the Guinea case.\n",
    "* Content-wise, however, there are some missing totals while the results for individual regions (columns other than `National`) are available. As a result, we opted to collect both the national totals and the sum of the individual regions, and to combine them by taking the maximum of both. As in the Liberia case, taking the maximum is an arbitrary choice and we could have opted to combine them differently instead (e.g., `min`, `average`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Read in Sierra Leone CSV\n",
    "# Note: unlike the other two countries, we read all columns because we need to sum them up to fill in missing data\n",
    "sl_df = pd.concat((pd.read_csv(f, parse_dates=['date']) for f in glob.glob(DATA_FOLDER+'/ebola/sl_data/*.csv')))\n",
    "\n",
    "# Collect cases\n",
    "df1 = sl_df[['date', 'variable', 'National']].copy().set_index('date')\n",
    "df1 = df1.loc[df1['variable'].isin(['cum_suspected', 'cum_probable', 'cum_confirmed'])][['National']].apply(pd.to_numeric, errors='coerce').groupby('date').sum()\n",
    "df1.columns = ['value']\n",
    "\n",
    "df2 = sl_df.copy().set_index('date')\n",
    "df2 = df2.loc[df2['variable'].isin(['cum_suspected', 'cum_probable', 'cum_confirmed'])]\n",
    "df2 = df2.loc[:, ~df2.columns.isin(['variable', 'National'])].fillna(0).apply(pd.to_numeric, errors='coerce')\n",
    "df2 = df2.groupby('date').sum()\n",
    "df2 = df2.sum(axis=1).to_frame('value')\n",
    "\n",
    "sl_cases = pd.concat([df1, df2]).sort_index().groupby('date').max().dropna()\n",
    "sl_cases\n",
    "\n",
    "# Collect deaths\n",
    "df1 = sl_df[['date', 'variable', 'National']].copy().set_index('date')\n",
    "df1 = df1.loc[df1['variable'].isin(['death_suspected', 'death_probable', 'death_confirmed'])][['National']].apply(pd.to_numeric, errors='coerce').groupby('date').sum()\n",
    "df1.columns = ['value']\n",
    "\n",
    "df2 = sl_df.copy().set_index('date')\n",
    "df2 = df2.loc[df2['variable'].isin(['death_suspected', 'death_probable', 'death_confirmed'])]\n",
    "df2 = df2.loc[:, ~df2.columns.isin(['variable', 'National'])].dropna().apply(pd.to_numeric, errors='coerce')\n",
    "df2 = df2.groupby('date').sum()\n",
    "df2 = df2.sum(axis=1).to_frame('value')\n",
    "\n",
    "sl_deaths = pd.concat([df1, df2]).sort_index().groupby('date').max().dropna()\n",
    "\n",
    "# Clean up the two frames\n",
    "sl_cases_cum = clean_frame(sl_cases)\n",
    "sl_deaths_cum = clean_frame(sl_deaths)\n",
    "\n",
    "# Convert to daily\n",
    "sl_cases_daily = cum_to_daily(sl_cases).to_frame('cases')\n",
    "sl_deaths_daily =  cum_to_daily(sl_deaths).to_frame('deaths')\n",
    "\n",
    "# Combine the two into a single frame\n",
    "sl = country_frame(sl_cases_daily, sl_deaths_daily, 'Sierra Leone')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have all three dataframes in a clean and structured format, we simply combine them into a single frame named `ebola`. Using that dataframe, we compute the daily averages per month as follows:\n",
    "1. Compute monthly totals, i.e. sum up all values for both new cases and deaths per month.\n",
    "2. Compute daily averages from there, i.e. divide the monthly totals by the corresponding number of days.\n",
    "\n",
    "The final result is presented in the `ebola_daily_averages` dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>cases</th>\n",
       "      <th>deaths</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>country</th>\n",
       "      <th>year</th>\n",
       "      <th>month</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Guinea</th>\n",
       "      <th rowspan=\"3\" valign=\"top\">2014</th>\n",
       "      <th>August</th>\n",
       "      <td>24.161290</td>\n",
       "      <td>15.774194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>September</th>\n",
       "      <td>14.000000</td>\n",
       "      <td>7.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>October</th>\n",
       "      <td>0.967742</td>\n",
       "      <td>0.806452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">Liberia</th>\n",
       "      <th rowspan=\"7\" valign=\"top\">2014</th>\n",
       "      <th>June</th>\n",
       "      <td>3.000000</td>\n",
       "      <td>1.633333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>July</th>\n",
       "      <td>7.000000</td>\n",
       "      <td>3.096774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>August</th>\n",
       "      <td>41.129032</td>\n",
       "      <td>23.870968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>September</th>\n",
       "      <td>74.800000</td>\n",
       "      <td>39.100000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>October</th>\n",
       "      <td>87.064516</td>\n",
       "      <td>20.612903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>November</th>\n",
       "      <td>37.600000</td>\n",
       "      <td>15.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>December</th>\n",
       "      <td>4.645161</td>\n",
       "      <td>4.290323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Sierra Leone</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">2014</th>\n",
       "      <th>August</th>\n",
       "      <td>38.387097</td>\n",
       "      <td>14.032258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>September</th>\n",
       "      <td>41.733333</td>\n",
       "      <td>5.433333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>October</th>\n",
       "      <td>69.967742</td>\n",
       "      <td>27.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>November</th>\n",
       "      <td>83.266667</td>\n",
       "      <td>14.233333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>December</th>\n",
       "      <td>37.548387</td>\n",
       "      <td>6.838710</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 cases     deaths\n",
       "country      year month                          \n",
       "Guinea       2014 August     24.161290  15.774194\n",
       "                  September  14.000000   7.500000\n",
       "                  October     0.967742   0.806452\n",
       "Liberia      2014 June        3.000000   1.633333\n",
       "                  July        7.000000   3.096774\n",
       "                  August     41.129032  23.870968\n",
       "                  September  74.800000  39.100000\n",
       "                  October    87.064516  20.612903\n",
       "                  November   37.600000  15.333333\n",
       "                  December    4.645161   4.290323\n",
       "Sierra Leone 2014 August     38.387097  14.032258\n",
       "                  September  41.733333   5.433333\n",
       "                  October    69.967742  27.000000\n",
       "                  November   83.266667  14.233333\n",
       "                  December   37.548387   6.838710"
      ]
     },
     "execution_count": 369,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Combine all 3 frames\n",
    "ebola = pd.concat([guinea, liberia, sl])\n",
    "\n",
    "# Compute monthly totals\n",
    "ebola_month_totals = ebola.reset_index().assign(year=lambda x: x['date'].dt.year, month=lambda x: x['date'].dt.month).groupby(by=['country', 'year', 'month']).sum()\n",
    "\n",
    "# Create daily averages per month\n",
    "ebola_daily_averages = ebola_month_totals.copy()\n",
    "ebola_daily_averages['cases'] = ebola_daily_averages['cases'] / ebola_daily_averages.index.map(lambda x: calendar.monthrange(x[1], x[2])[1])\n",
    "ebola_daily_averages['deaths'] = ebola_daily_averages['deaths'] / ebola_daily_averages.index.map(lambda x: calendar.monthrange(x[1], x[2])[1])\n",
    "\n",
    "# We are done! ebola_daily_averages contains the requested info\n",
    "\n",
    "# Prettifying months (month name instead of number)\n",
    "ebola_daily_averages = ebola_daily_averages.reset_index()\n",
    "ebola_daily_averages['month'] = ebola_daily_averages['month'].map(lambda x: calendar.month_name[x])\n",
    "ebola_daily_averages = ebola_daily_averages.set_index(['country', 'year', 'month'])\n",
    "ebola_daily_averages"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Methodology\n",
    "\n",
    "Looking at the files with little understanding of RNA seequencing, it appears that the 9 spreadsheets contain tuples of the form $(taxonomy, value)$ while the metadata contains tuples of $(barcode, group, sample)$. It appears the $barcode$ corresponds to the name of a spreadsheet.\n",
    "\n",
    "We therefore assume that the final output frame should contain tuples of the form $(barcode, group, sample, taxonomy, value)$ where $barcode$ is joined with the data contained in the corresponding spreadsheet. It appears natural to index the data using the following multi-index: $(barcode, group, sample, taxonomy)$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 422,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th>group</th>\n",
       "      <th>sample</th>\n",
       "      <th>taxon</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">MID1</th>\n",
       "      <th rowspan=\"30\" valign=\"top\">EXTRACTION CONTROL</th>\n",
       "      <th rowspan=\"30\" valign=\"top\">unknown</th>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methanosarcinaceae Methanimicrococcus</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methermicoccaceae Methermicoccus</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglobales Archaeoglobaceae Ferroglobus</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglobales Archaeoglobaceae Geoglobus</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haloplanus</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haloquadratum</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Halosimplex</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Natronobacterium</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Natronomonas</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Methanococci Methanococcales Methanocaldococcaceae Methanotorris</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Methanopyri Methanopyrales Methanopyraceae Methanopyrus</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Thermoplasmata Thermoplasmatales Thermoplasmatales_incertae_sedis Thermogymnomonas</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Acidothermaceae Acidothermus</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Beutenbergiaceae Salana</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Bogoriellaceae Bogoriella</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Cellulomonadaceae Tropheryma</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Corynebacteriaceae Turicella</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Cryptosporangiaceae Cryptosporangium</th>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Dermabacteraceae Dermabacter</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Dermabacteraceae Devriesea</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Dermatophilaceae Kineosphaera</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Frankineae_incertae_sedis Fodinicola</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Geodermatophilaceae Blastococcus</th>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Geodermatophilaceae Geodermatophilus</th>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Intrasporangiaceae Kribbia</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">MID9</th>\n",
       "      <th rowspan=\"30\" valign=\"top\">Control 2</th>\n",
       "      <th rowspan=\"30\" valign=\"top\">stool</th>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Oceanospirillales Oceanospirillaceae Neptuniibacter</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Oceanospirillales Oceanospirillales_incertae_sedis Spongiispira</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Oceanospirillales Oleiphilaceae Oleiphilus</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pasteurellales Pasteurellaceae Haemophilus</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Moraxellaceae Alkanindiges</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Moraxellaceae Perlucidibaca</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Azomonas</th>\n",
       "      <td>805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Azorhizophilus</th>\n",
       "      <td>862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Azotobacter</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Pseudomonas</th>\n",
       "      <td>3534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Piscirickettsiaceae Cycloclasticus</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Piscirickettsiaceae Sulfurivirga</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Thiotrichaceae Beggiatoa</th>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Thiotrichales_incertae_sedis Fangia</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Sinobacteraceae Alkanibacter</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Sinobacteraceae Sinobacter</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Aquimonas</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Frateuria</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Ignatzschineria</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Luteibacter</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Synergistetes\" Synergistia Synergistales Synergistaceae Aminomonas</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Tenericutes\" Mollicutes Anaeroplasmatales Anaeroplasmataceae Asteroleplasma</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Tenericutes\" Mollicutes Haloplasmatales Haloplasmataceae Haloplasma</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermodesulfobacteria\" Thermodesulfobacteria Thermodesulfobacteriales Thermodesulfobacteriaceae Caldimicrobium</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermodesulfobacteria\" Thermodesulfobacteria Thermodesulfobacteriales Thermodesulfobacteriaceae Thermodesulfatator</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermotogae\" Thermotogae Thermotogales Thermotogaceae Geotoga</th>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermotogae\" Thermotogae Thermotogales Thermotogaceae Kosmotoga</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Opitutales Opitutaceae Opitutus</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Chlorarachniophyceae</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Streptophyta</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2396 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                       value\n",
       "barcode group              sample  taxon                                                    \n",
       "MID1    EXTRACTION CONTROL unknown Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      7\n",
       "                                   Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      2\n",
       "                                   Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...      3\n",
       "                                   Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...      3\n",
       "                                   Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...      7\n",
       "                                   Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...      1\n",
       "                                   Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...      1\n",
       "                                   Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...      1\n",
       "                                   Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...      1\n",
       "                                   Archaea \"Euryarchaeota\" Halobacteria Halobacter...      1\n",
       "                                   Archaea \"Euryarchaeota\" Halobacteria Halobacter...      4\n",
       "                                   Archaea \"Euryarchaeota\" Halobacteria Halobacter...      1\n",
       "                                   Archaea \"Euryarchaeota\" Halobacteria Halobacter...      2\n",
       "                                   Archaea \"Euryarchaeota\" Halobacteria Halobacter...      4\n",
       "                                   Archaea \"Euryarchaeota\" Methanococci Methanococ...      1\n",
       "                                   Archaea \"Euryarchaeota\" Methanopyri Methanopyra...     12\n",
       "                                   Archaea \"Euryarchaeota\" Thermoplasmata Thermopl...      2\n",
       "                                   Bacteria \"Actinobacteria\" Actinobacteria Actino...      2\n",
       "                                   Bacteria \"Actinobacteria\" Actinobacteria Actino...      1\n",
       "                                   Bacteria \"Actinobacteria\" Actinobacteria Actino...      1\n",
       "                                   Bacteria \"Actinobacteria\" Actinobacteria Actino...      1\n",
       "                                   Bacteria \"Actinobacteria\" Actinobacteria Actino...      2\n",
       "                                   Bacteria \"Actinobacteria\" Actinobacteria Actino...     53\n",
       "                                   Bacteria \"Actinobacteria\" Actinobacteria Actino...      1\n",
       "                                   Bacteria \"Actinobacteria\" Actinobacteria Actino...      2\n",
       "                                   Bacteria \"Actinobacteria\" Actinobacteria Actino...      3\n",
       "                                   Bacteria \"Actinobacteria\" Actinobacteria Actino...      1\n",
       "                                   Bacteria \"Actinobacteria\" Actinobacteria Actino...     99\n",
       "                                   Bacteria \"Actinobacteria\" Actinobacteria Actino...     17\n",
       "                                   Bacteria \"Actinobacteria\" Actinobacteria Actino...      2\n",
       "...                                                                                      ...\n",
       "MID9    Control 2          stool   Bacteria \"Proteobacteria\" Gammaproteobacteria O...     10\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria O...      1\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria O...      1\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria P...      1\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria P...     12\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria P...      4\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria P...    805\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria P...    862\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria P...     12\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria P...   3534\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria T...      4\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria T...      1\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria T...     25\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria T...      2\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria X...      3\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria X...      2\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria X...      1\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria X...      3\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria X...      2\n",
       "                                   Bacteria \"Proteobacteria\" Gammaproteobacteria X...      1\n",
       "                                   Bacteria \"Synergistetes\" Synergistia Synergista...      3\n",
       "                                   Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...      1\n",
       "                                   Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...     11\n",
       "                                   Bacteria \"Thermodesulfobacteria\" Thermodesulfob...      1\n",
       "                                   Bacteria \"Thermodesulfobacteria\" Thermodesulfob...      1\n",
       "                                   Bacteria \"Thermotogae\" Thermotogae Thermotogale...     43\n",
       "                                   Bacteria \"Thermotogae\" Thermotogae Thermotogale...     16\n",
       "                                   Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...      1\n",
       "                                   Bacteria Cyanobacteria Cyanobacteria  Chloropla...      1\n",
       "                                   Bacteria Cyanobacteria Cyanobacteria  Chloropla...     10\n",
       "\n",
       "[2396 rows x 1 columns]"
      ]
     },
     "execution_count": 422,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read each spreadsheet into a dataframe and append a column for the barcode\n",
    "mid = []\n",
    "for i in range(1,10):\n",
    "    m = pd.read_excel(DATA_FOLDER+'/microbiome/MID%d.xls' % i, names=['taxon', 'value'], sheetname='Sheet 1', header=None)\n",
    "    m['barcode'] = 'MID%d' % i\n",
    "    mid.append(m)\n",
    "\n",
    "# Put all spreadsheets into a single dataframe\n",
    "df = pd.concat(mid)\n",
    "\n",
    "# Read metadata\n",
    "meta = pd.read_excel(DATA_FOLDER+'/microbiome/metadata.xls', sheetname='Sheet1')\n",
    "meta.columns = map(str.lower, meta.columns) # clean up column names\n",
    "\n",
    "# Merge with metadata (will join on barcode)\n",
    "mb = pd.merge(df, meta)\n",
    "\n",
    "# Replace NaNs with unknown and set index\n",
    "mb[['sample']] = mb[['sample']].replace(np.nan, 'unknown')\n",
    "mb = mb.set_index(['barcode', 'group', 'sample', 'taxon'])\n",
    "\n",
    "# Make sure we have a unique index and no NaNs\n",
    "assert(mb.index.is_unique)\n",
    "assert(len(mb[mb.isnull().any(axis=1)]) == 0)\n",
    "\n",
    "# Here goes the microbiome dataframe\n",
    "mb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<hr><h2>Data frame:titanic3</h2>1309 observations and 14 variables, maximum # NAs:1188<hr>\n",
       "<TABLE BORDER>\n",
       "<tr><td>Name</td><td>Labels</td><td>Units</td><td>Levels</td><td>Storage</td><td>NAs</td></tr>\n",
       "<tr><td>pclass</td><td></td><td></td><td><a href=\"#pclass\"><div align=right>  3</div></a></td><td>integer</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>survived</td><td>Survived</td><td></td><td><div align=right></div></td><td>double</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>name</td><td>Name</td><td></td><td><div align=right></div></td><td>character</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>sex</td><td></td><td></td><td><a href=\"#sex\"><div align=right>  2</div></a></td><td>integer</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>age</td><td>Age</td><td>Year</td><td><div align=right></div></td><td>double</td><td><div align=right> 263</div></td></tr>\n",
       "<tr><td>sibsp</td><td>Number of Siblings/Spouses Aboard</td><td></td><td><div align=right></div></td><td>double</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>parch</td><td>Number of Parents/Children Aboard</td><td></td><td><div align=right></div></td><td>double</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>ticket</td><td>Ticket Number</td><td></td><td><div align=right></div></td><td>character</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>fare</td><td>Passenger Fare</td><td>British Pound (\\243)</td><td><div align=right></div></td><td>double</td><td><div align=right>   1</div></td></tr>\n",
       "<tr><td>cabin</td><td></td><td></td><td><a href=\"#cabin\"><div align=right>187</div></a></td><td>integer</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>embarked</td><td></td><td></td><td><a href=\"#embarked\"><div align=right>  3</div></a></td><td>integer</td><td><div align=right>   2</div></td></tr>\n",
       "<tr><td>boat</td><td></td><td></td><td><a href=\"#boat\"><div align=right> 28</div></a></td><td>integer</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>body</td><td>Body Identification Number</td><td></td><td><div align=right></div></td><td>double</td><td><div align=right>1188</div></td></tr>\n",
       "<tr><td>home.dest</td><td>Home/Destination</td><td></td><td><div align=right></div></td><td>character</td><td><div align=right>   0</div></td></tr>\n",
       "</TABLE>\n",
       "<hr>\n",
       "<TABLE BORDER>\n",
       "<tr><td>Variable</td><td>Levels</td></tr>\n",
       "<tr><td><a name=\"pclass\">pclass</a></td><td>1st</td></tr>\n",
       "<tr><td></td><td>2nd</td></tr>\n",
       "<tr><td></td><td>3rd</td></tr>\n",
       "<tr><td><a name=\"sex\">sex</a></td><td>female</td></tr>\n",
       "<tr><td></td><td>male</td></tr>\n",
       "<tr><td><a name=\"cabin\">cabin</a></td><td></td></tr>\n",
       "<tr><td></td><td>A10</td></tr>\n",
       "<tr><td></td><td>A11</td></tr>\n",
       "<tr><td></td><td>A14</td></tr>\n",
       "<tr><td></td><td>A16</td></tr>\n",
       "<tr><td></td><td>A18</td></tr>\n",
       "<tr><td></td><td>A19</td></tr>\n",
       "<tr><td></td><td>A20</td></tr>\n",
       "<tr><td></td><td>A21</td></tr>\n",
       "<tr><td></td><td>A23</td></tr>\n",
       "<tr><td></td><td>A24</td></tr>\n",
       "<tr><td></td><td>A26</td></tr>\n",
       "<tr><td></td><td>A29</td></tr>\n",
       "<tr><td></td><td>A31</td></tr>\n",
       "<tr><td></td><td>A32</td></tr>\n",
       "<tr><td></td><td>A34</td></tr>\n",
       "<tr><td></td><td>A36</td></tr>\n",
       "<tr><td></td><td>A5</td></tr>\n",
       "<tr><td></td><td>A6</td></tr>\n",
       "<tr><td></td><td>A7</td></tr>\n",
       "<tr><td></td><td>A9</td></tr>\n",
       "<tr><td></td><td>B10</td></tr>\n",
       "<tr><td></td><td>B101</td></tr>\n",
       "<tr><td></td><td>B102</td></tr>\n",
       "<tr><td></td><td>B11</td></tr>\n",
       "<tr><td></td><td>B18</td></tr>\n",
       "<tr><td></td><td>B19</td></tr>\n",
       "<tr><td></td><td>B20</td></tr>\n",
       "<tr><td></td><td>B22</td></tr>\n",
       "<tr><td></td><td>B24</td></tr>\n",
       "<tr><td></td><td>B26</td></tr>\n",
       "<tr><td></td><td>B28</td></tr>\n",
       "<tr><td></td><td>B3</td></tr>\n",
       "<tr><td></td><td>B30</td></tr>\n",
       "<tr><td></td><td>B35</td></tr>\n",
       "<tr><td></td><td>B36</td></tr>\n",
       "<tr><td></td><td>B37</td></tr>\n",
       "<tr><td></td><td>B38</td></tr>\n",
       "<tr><td></td><td>B39</td></tr>\n",
       "<tr><td></td><td>B4</td></tr>\n",
       "<tr><td></td><td>B41</td></tr>\n",
       "<tr><td></td><td>B42</td></tr>\n",
       "<tr><td></td><td>B45</td></tr>\n",
       "<tr><td></td><td>B49</td></tr>\n",
       "<tr><td></td><td>B5</td></tr>\n",
       "<tr><td></td><td>B50</td></tr>\n",
       "<tr><td></td><td>B51 B53 B55</td></tr>\n",
       "<tr><td></td><td>B52 B54 B56</td></tr>\n",
       "<tr><td></td><td>B57 B59 B63 B66</td></tr>\n",
       "<tr><td></td><td>B58 B60</td></tr>\n",
       "<tr><td></td><td>B61</td></tr>\n",
       "<tr><td></td><td>B69</td></tr>\n",
       "<tr><td></td><td>B71</td></tr>\n",
       "<tr><td></td><td>B73</td></tr>\n",
       "<tr><td></td><td>B77</td></tr>\n",
       "<tr><td></td><td>B78</td></tr>\n",
       "<tr><td></td><td>B79</td></tr>\n",
       "<tr><td></td><td>B80</td></tr>\n",
       "<tr><td></td><td>B82 B84</td></tr>\n",
       "<tr><td></td><td>B86</td></tr>\n",
       "<tr><td></td><td>B94</td></tr>\n",
       "<tr><td></td><td>B96 B98</td></tr>\n",
       "<tr><td></td><td>C101</td></tr>\n",
       "<tr><td></td><td>C103</td></tr>\n",
       "<tr><td></td><td>C104</td></tr>\n",
       "<tr><td></td><td>C105</td></tr>\n",
       "<tr><td></td><td>C106</td></tr>\n",
       "<tr><td></td><td>C110</td></tr>\n",
       "<tr><td></td><td>C111</td></tr>\n",
       "<tr><td></td><td>C116</td></tr>\n",
       "<tr><td></td><td>C118</td></tr>\n",
       "<tr><td></td><td>C123</td></tr>\n",
       "<tr><td></td><td>C124</td></tr>\n",
       "<tr><td></td><td>C125</td></tr>\n",
       "<tr><td></td><td>C126</td></tr>\n",
       "<tr><td></td><td>C128</td></tr>\n",
       "<tr><td></td><td>C130</td></tr>\n",
       "<tr><td></td><td>C132</td></tr>\n",
       "<tr><td></td><td>C148</td></tr>\n",
       "<tr><td></td><td>C2</td></tr>\n",
       "<tr><td></td><td>C22 C26</td></tr>\n",
       "<tr><td></td><td>C23 C25 C27</td></tr>\n",
       "<tr><td></td><td>C28</td></tr>\n",
       "<tr><td></td><td>C30</td></tr>\n",
       "<tr><td></td><td>C31</td></tr>\n",
       "<tr><td></td><td>C32</td></tr>\n",
       "<tr><td></td><td>C39</td></tr>\n",
       "<tr><td></td><td>C45</td></tr>\n",
       "<tr><td></td><td>C46</td></tr>\n",
       "<tr><td></td><td>C47</td></tr>\n",
       "<tr><td></td><td>C49</td></tr>\n",
       "<tr><td></td><td>C50</td></tr>\n",
       "<tr><td></td><td>C51</td></tr>\n",
       "<tr><td></td><td>C52</td></tr>\n",
       "<tr><td></td><td>C53</td></tr>\n",
       "<tr><td></td><td>C54</td></tr>\n",
       "<tr><td></td><td>C55 C57</td></tr>\n",
       "<tr><td></td><td>C6</td></tr>\n",
       "<tr><td></td><td>C62 C64</td></tr>\n",
       "<tr><td></td><td>C65</td></tr>\n",
       "<tr><td></td><td>C68</td></tr>\n",
       "<tr><td></td><td>C7</td></tr>\n",
       "<tr><td></td><td>C70</td></tr>\n",
       "<tr><td></td><td>C78</td></tr>\n",
       "<tr><td></td><td>C80</td></tr>\n",
       "<tr><td></td><td>C82</td></tr>\n",
       "<tr><td></td><td>C83</td></tr>\n",
       "<tr><td></td><td>C85</td></tr>\n",
       "<tr><td></td><td>C86</td></tr>\n",
       "<tr><td></td><td>C87</td></tr>\n",
       "<tr><td></td><td>C89</td></tr>\n",
       "<tr><td></td><td>C90</td></tr>\n",
       "<tr><td></td><td>C91</td></tr>\n",
       "<tr><td></td><td>C92</td></tr>\n",
       "<tr><td></td><td>C93</td></tr>\n",
       "<tr><td></td><td>C95</td></tr>\n",
       "<tr><td></td><td>C97</td></tr>\n",
       "<tr><td></td><td>C99</td></tr>\n",
       "<tr><td></td><td>D</td></tr>\n",
       "<tr><td></td><td>D10 D12</td></tr>\n",
       "<tr><td></td><td>D11</td></tr>\n",
       "<tr><td></td><td>D15</td></tr>\n",
       "<tr><td></td><td>D17</td></tr>\n",
       "<tr><td></td><td>D19</td></tr>\n",
       "<tr><td></td><td>D20</td></tr>\n",
       "<tr><td></td><td>D21</td></tr>\n",
       "<tr><td></td><td>D22</td></tr>\n",
       "<tr><td></td><td>D26</td></tr>\n",
       "<tr><td></td><td>D28</td></tr>\n",
       "<tr><td></td><td>D30</td></tr>\n",
       "<tr><td></td><td>D33</td></tr>\n",
       "<tr><td></td><td>D34</td></tr>\n",
       "<tr><td></td><td>D35</td></tr>\n",
       "<tr><td></td><td>D36</td></tr>\n",
       "<tr><td></td><td>D37</td></tr>\n",
       "<tr><td></td><td>D38</td></tr>\n",
       "<tr><td></td><td>D40</td></tr>\n",
       "<tr><td></td><td>D43</td></tr>\n",
       "<tr><td></td><td>D45</td></tr>\n",
       "<tr><td></td><td>D46</td></tr>\n",
       "<tr><td></td><td>D47</td></tr>\n",
       "<tr><td></td><td>D48</td></tr>\n",
       "<tr><td></td><td>D49</td></tr>\n",
       "<tr><td></td><td>D50</td></tr>\n",
       "<tr><td></td><td>D56</td></tr>\n",
       "<tr><td></td><td>D6</td></tr>\n",
       "<tr><td></td><td>D7</td></tr>\n",
       "<tr><td></td><td>D9</td></tr>\n",
       "<tr><td></td><td>E10</td></tr>\n",
       "<tr><td></td><td>E101</td></tr>\n",
       "<tr><td></td><td>E12</td></tr>\n",
       "<tr><td></td><td>E121</td></tr>\n",
       "<tr><td></td><td>E17</td></tr>\n",
       "<tr><td></td><td>E24</td></tr>\n",
       "<tr><td></td><td>E25</td></tr>\n",
       "<tr><td></td><td>E31</td></tr>\n",
       "<tr><td></td><td>E33</td></tr>\n",
       "<tr><td></td><td>E34</td></tr>\n",
       "<tr><td></td><td>E36</td></tr>\n",
       "<tr><td></td><td>E38</td></tr>\n",
       "<tr><td></td><td>E39 E41</td></tr>\n",
       "<tr><td></td><td>E40</td></tr>\n",
       "<tr><td></td><td>E44</td></tr>\n",
       "<tr><td></td><td>E45</td></tr>\n",
       "<tr><td></td><td>E46</td></tr>\n",
       "<tr><td></td><td>E49</td></tr>\n",
       "<tr><td></td><td>E50</td></tr>\n",
       "<tr><td></td><td>E52</td></tr>\n",
       "<tr><td></td><td>E58</td></tr>\n",
       "<tr><td></td><td>E60</td></tr>\n",
       "<tr><td></td><td>E63</td></tr>\n",
       "<tr><td></td><td>E67</td></tr>\n",
       "<tr><td></td><td>E68</td></tr>\n",
       "<tr><td></td><td>E77</td></tr>\n",
       "<tr><td></td><td>E8</td></tr>\n",
       "<tr><td></td><td>F</td></tr>\n",
       "<tr><td></td><td>F E46</td></tr>\n",
       "<tr><td></td><td>F E57</td></tr>\n",
       "<tr><td></td><td>F E69</td></tr>\n",
       "<tr><td></td><td>F G63</td></tr>\n",
       "<tr><td></td><td>F G73</td></tr>\n",
       "<tr><td></td><td>F2</td></tr>\n",
       "<tr><td></td><td>F33</td></tr>\n",
       "<tr><td></td><td>F38</td></tr>\n",
       "<tr><td></td><td>F4</td></tr>\n",
       "<tr><td></td><td>G6</td></tr>\n",
       "<tr><td></td><td>T</td></tr>\n",
       "<tr><td><a name=\"embarked\">embarked</a></td><td>Cherbourg</td></tr>\n",
       "<tr><td></td><td>Queenstown</td></tr>\n",
       "<tr><td></td><td>Southampton</td></tr>\n",
       "<tr><td><a name=\"boat\">boat</a></td><td></td></tr>\n",
       "<tr><td></td><td>1</td></tr>\n",
       "<tr><td></td><td>10</td></tr>\n",
       "<tr><td></td><td>11</td></tr>\n",
       "<tr><td></td><td>12</td></tr>\n",
       "<tr><td></td><td>13</td></tr>\n",
       "<tr><td></td><td>13 15</td></tr>\n",
       "<tr><td></td><td>13 15 B</td></tr>\n",
       "<tr><td></td><td>14</td></tr>\n",
       "<tr><td></td><td>15</td></tr>\n",
       "<tr><td></td><td>15 16</td></tr>\n",
       "<tr><td></td><td>16</td></tr>\n",
       "<tr><td></td><td>2</td></tr>\n",
       "<tr><td></td><td>3</td></tr>\n",
       "<tr><td></td><td>4</td></tr>\n",
       "<tr><td></td><td>5</td></tr>\n",
       "<tr><td></td><td>5 7</td></tr>\n",
       "<tr><td></td><td>5 9</td></tr>\n",
       "<tr><td></td><td>6</td></tr>\n",
       "<tr><td></td><td>7</td></tr>\n",
       "<tr><td></td><td>8</td></tr>\n",
       "<tr><td></td><td>8 10</td></tr>\n",
       "<tr><td></td><td>9</td></tr>\n",
       "<tr><td></td><td>A</td></tr>\n",
       "<tr><td></td><td>B</td></tr>\n",
       "<tr><td></td><td>C</td></tr>\n",
       "<tr><td></td><td>C D</td></tr>\n",
       "<tr><td></td><td>D</td></tr>\n",
       "</TABLE>\n",
       "<hr>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
